Use conda-lock for reproducible conda env

tobiasdiez opened the pull request:
<!-- ^^^^^
Please provide a concise, informative and self-explanatory title.
Don't put issue numbers in there, do this in the PR body below.
For example, instead of "Fixes #1234" use "Introduce new method to calculate 1+1"
-->
<!-- Describe your changes here in detail -->

<!-- Why is this change required? What problem does it solve? -->
<!-- If this PR resolves an open issue, please link to it here. For example "Fixes #12345". -->
<!-- If your change requires a documentation PR, please link it appropriately. -->

The conda workflow is broken since a few months due to various issues that arise from some packages being updated in conda-forge and sage not being quite compatible with the new version. For example, currently it fails as Cython 3.0.3 is used.

To prevent such issues happening in the feature, we use [conda-lock](https://conda.github.io/conda-lock/) to create a lock file for each os + python combination, from which one can then create a reproducible conda env. These lock files are created using commands of the form:
```bash
conda-lock --channel conda-forge --kind env --platform linux-64 --platform osx-64 --file ./src/environment-dev-3.10.yml --filename-template "src/environment-dev-3.10-{platform}"
```


Follow-up: Add a github workflow that updates these lock files automatically, similar to:
- https://github.com/ibis-project/ibis/blob/master/.github/workflows/conda-lock.yml
- https://github.com/svopper/munaiah-analyser/blob/main/out/SciTools/iris/.github_workflows_refresh-lockfiles.yml
- https://github.com/chipsalliance/fpga-interchange-tests/blob/main/.github/workflows/update_conda_lock.yml
- https://github.com/ESMValGroup/ESMValTool/blob/main/.github/workflows/create-condalock-file.yml

### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!-- Feel free to remove irrelevant items. -->

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### :hourglass: Dependencies

- https://github.com/sagemath/sage/pull/36405: to get env files for each python version
- #36767 

<!-- List all open PRs that this PR logically depends on
- #12345: short description why this is a dependency
- #34567: ...
-->

<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

tobiasdiez committed:
Use conda-lock for reproducible conda env

isuruf commented on src/pyproject.toml.m4:
```diff
@@ -21,3 +21,8 @@ requires = [
+
+[tool.conda-lock]
+platforms = [
+    'osx-64', 'linux-64'
```
```suggestion
    'osx-64', 'linux-64', 'linux-aarch64', 'osx-arm64'
```

tobiasdiez replied:
Good point, thanks!

tobiasdiez committed:
fix env file name

tobiasdiez committed:
add more lock files

tobiasdiez committed:
Add more supported architectures

Co-authored-by: Isuru Fernando <isuruf@gmail.com>

tobiasdiez committed:
fix lock files

mkoeppe commented on src/doc/en/installation/conda.rst:
```diff
@@ -68,18 +68,15 @@ Using conda to provide system packages for the Sage distribution
-    ``environment.yml`` to create an environment with all standard and optional
+    If you use a different architecture, replace ``linux`` by ``macos``.
+    Alternatively, use ``environment-optional-3.11-linux.yml`` in place of
+    ``environment-3.11-linux.yml`` to create an environment with all standard and optional
```
I think we cannot completely eliminate mention of the files `environment-3.9.yml` etc. and the `bootstrap-conda` script. 

At least in the packaging section of the developer's guide it should be explained how to maintain the locked files, and who should do this when.

tobiasdiez replied:
My idea was to push the maintenance of the lock files to a github action.

mkoeppe replied:
It needs to be maintainable by developers, so a bit of documentation is necessary.

tobiasdiez replied:
I don't agree

tobiasdiez replied:
Still don't really see the point, but done.

mkoeppe commented on src/pyproject.toml.m4:
```diff
@@ -21,3 +21,8 @@ requires = [
+[tool.conda-lock]
+platforms = [
+    'osx-64', 'linux-64', 'linux-aarch64', 'osx-arm64'
+]
```
OK, the config for aarch/arm is here but shouldn't the resulting files be committed too?

tobiasdiez replied:
Yes, I would prefer to do this as part of the follow-up that adds the automatic update. It's a bit of tedious work to create all of them manually.

mkoeppe replied:
How about adding a maintenance script that makes the necessary calls. 

The GH Action can then call that script.

tobiasdiez replied:
Done.

Commit:
src/doc/en/installation/conda.rst: Update for versioned environment-... files

Commit:
bootstrap, Makefile (bootstrap-clean): Update for versioned environment-... files

Commit:
tox.ini (local-conda-environment): Update for versioned environment-... files

Commit:
.github/workflows/ci-conda.yml: Update for versioned environment*.yml

Commit:
tox.ini (conda-environment-{src,dev}), .github/workflows/ci-conda.yml: Use configure --enable-system-site-packages

tobiasdiez committed:
Merge branch 'develop' into conda-lock

Commit:
.github/workflows/ci-conda.yml: Fix environment name

Commit:
bootstrap-conda: Generate environment files with hard-coded Python minor versions (environment-3.9.yml etc.)

This is convenient for conda-lock.

Partial cherry-pick from "Use conda-lock for reproducible conda env"

Commit:
bootstrap-conda: Refactor using sage-get-system-packages

tobiasdiez committed:
Merge remote-tracking branch 'origin/develop' into conda-lock

tobiasdiez committed:
Merge remote-tracking branch 'mkoeppe/bootstrap_conda_refactor' into conda-lock

tobiasdiez committed:
update lock files

tobiasdiez committed:
exclude 3.0.3

tobiasdiez committed:
update lock files

tobiasdiez committed:
bootstrap-conda: Only one 'python' line + speed up

tobiasdiez committed:
add lock files for aarch and arm

tobiasdiez committed:
convert update script to python

tobiasdiez committed:
Merge remote-tracking branch 'upstream/develop' into conda-lock

tobiasdiez committed:
Exclude unnecessary prereq

tobiasdiez committed:
update arm names

tobiasdiez committed:
readd python to env

tobiasdiez committed:
correct print output

tobiasdiez committed:
update lock files

dimpase commented on build/pkgs/_prereq/distros/conda.txt:
```diff
@@ -1,7 +1,8 @@
 python
-tar
-bc
+# Not needed on conda
```
I don't understand this change. This is something used not only for CI, as far as I know.

dimpase replied:
it would only make sense to me if the packages listed there are always included with conda, but that's not the case AFAIK.

tobiasdiez replied:
It's a long time since I did this change. I don't quite remember why I removed them, there was some problem. Most of these (all?) are only prerequisites of sage-the-distro, and not sagelib-the-python-package. With conda, we don't use make or untar things. (Not so sure about perl, bc and compilers, but they don't seem to be needed/are included anyway)

dimpase replied:
> It's a long time since I did this change. I don't quite remember why I removed them, there was some problem. Most of these (all?) are only prerequisites of sage-the-distro, and not sagelib-the-python-package. With conda, we don't use make or untar things. (Not so sure about perl, bc and compilers, but they don't seem to be needed/are included anyway)

this is only true if one does not do conda based development, for which you will need compilers, etc etc.

tobiasdiez replied:
I don't quite understand what you mean by this. Could you please rephrase it?

Looking at the ci (eg https://github.com/sagemath/sage/actions/runs/7003785707/job/19050185446?pr=35986), one sees that the build + test is successful without bc and tar (c and fortran compilers, make, m4, perl are still pulled-in by other dependencies). But as I said above, I don't remember why I've removed these. Should I try to readd all of them, or are there indeed packages that are clearly not needed as actual prerequisites for sagelib?

mkoeppe replied:
Dima is right here, these packages cannot be removed. `build/pkgs/_prereq/distros/conda.txt` are the minimal requirements for building the Sage distribution from source; not just the Sage library. This is not tested in ci-conda, but it is tested using the CI Linux workflow. See https://github.com/sagemath/sage/actions/runs/7076264535/job/19259311074 -- note there is a build failure (needs #23712)

tobiasdiez replied:
I've removed these changes again. It would be good to clarify the purpose of `_prereq`, since it says `This dummy package represents the minimal requirements (system packages)
for installing SageMath from source.` which is not really what it does according to the discussion above, at least if you understand SageMath = sagelib. Most of these dependencies should then be added directly to sagelib, and then we can exclude `_prereq` from the env files under `src`. Let's leave this for another PR.

tobiasdiez committed:
revert changes to prereq

tobiasdiez committed:
Limit to older compilers to fix ci issues on macos

tobiasdiez committed:
fix name of lock file update script in docs

tobiasdiez committed:
install conda-lock in conda dev env

tobiasdiez committed:
rerender lock files

tobiasdiez committed:
Merge remote-tracking branch 'upstream/develop' into pr/tobiasdiez/35986

tobiasdiez committed:
add arb explicitly

tobiasdiez committed:
Remove arb again

tobiasdiez committed:
compilers is not a macos issue

tobiasdiez committed:
Merge remote-tracking branch 'tobiasdiez/conda_maxparallel' into conda-lock

tobiasdiez committed:
Merge remote-tracking branch 'upstream/develop' into conda-lock

tobiasdiez committed:
update conda lock files

tobiasdiez reviewed (commented):
Okay, done.

dimpase reviewed (approved):
OK

tobiasdiez committed:
Merge branch 'develop' into conda-lock
